### ISSUE TITLE
Git diffs use pagination inconsistently

### ISSUE THREAD

**Issue by ZimbiX** [Jguer/yay#1243]
#### Affected Version

yay v9.4.6 - libalpm v12.0.1

#### Issue

When I'm reviewing AUR package changes while running `yay`, the diffs are paginated inconsistently. If there's a package with a diff longer than the screen height, it'll show only that diff in what seems to be `less`. When I press Q to continue reviewing, and what remains is a bunch of packages with small diffs, they end up in the scrollback, and I have to scroll back up to see them all, trying to work out where I was up to.

It seems like the current functionality is something akin to sequentially running `git diff ...` in each package's repo, which has the effect of each one separately being possibly paginated.

I think the desired functionality would be to collect all the diffs, and then paginate that. This could be done by adding the `--no-pager` flag to the Git command (`git --no-pager diff`), and piping that into `less` (but then needing to use the `--color` Git flag to keep colour enabled, and `less -R` to interpret ANSI colour safely).

Also, including titles to separate the packages diffs more clearly would be great.

**Comment by ZimbiX** [comment]
Right, sorry, I didn't realise this was addressed in the readme:

> #### Sometimes diffs are printed to the terminal, and other times they are paged via less. How do I fix this?
> 
> Yay uses `git diff` to display diffs, which by default tells less not to
> page if the output can fit into one terminal length. This behavior can be
> overridden by exporting your own flags (`export LESS=SRX`).

Using this workaround, I think I prefer `LESS=SR`, which forces each package's diff to be separately paginated on a clean screen, no matter the length.

A quick improvement could be for Yay to default the `LESS` environment variable to `SR` if it's unset. Although this could make it tricky to quit early if you disagree with the content of a diff.

Still, I think it would be preferable to show all the diffs at once, separated by title.

For now at least, I will be adding this alias to my dotfiles:

```bash
alias yay='LESS=SR yay'
```

**Comment by HaleTom** [comment]
I think the option you're looking for is:
```
       -F or --quit-if-one-screen
              Causes less to automatically exit if the entire file can be displayed on the first screen.
```
If `less` were invoked with `-+F`, this would reset this to the default value. (I tested, it works on a simple file view)

So, maybe you could invoke `git` as `LESS="$LESS -+F" git diff ...`

Because invoking as `LESS="$LESS -+F" yay ...` sadly doesn't work.

Perhaps each call to git diff needs to be piped to `less -+F` rather than being printed directly to the screen?

**Comment by grandchild** [comment]
I think this issue could be closed? @ZimbiX  @Jguer

**Comment by HaleTom** [comment]
@Jguer @ZimbiX @grandchild Please consider reopening this issue...

The suggestions I've made don't resolve the issue.

Example: if there are 5 diffs that take up about 20 lines each, and these are printed one after another, all 100 lines will be dumped at once, and only the last screenful will actually be visible.

**Comment by HaleTom** [comment]
This works, but prints every file separately. It pauses after each file.

`PAGER='less $LESS -+F' /usr/bin/yay -aSu`

My preferred behaviour is printing the `diff` for the whole package, and pausing only once at the end.

**Comment by ZimbiX** [comment]
Please reopen

**Comment by afontenot** [comment]
> My preferred behaviour is printing the `diff` for the whole package, and pausing only once at the end.

I agree; this behavior would make a lot more sense. It's weird that it's not already the default because it's the most obvious way to diff an AUR repository. If you have an AUR repo cloned locally, and you do `git diff HEAD origin`, it just works, exactly like you'd expect. I think that's what many users would like to see yay do as well. 

@Jguer I believe this was closed by mistake. It's "stale", but only because there isn't a lot users can add to the topic.

**Comment by danielwerg** [comment]
this have bother me for way too long. here is my workaround with [delta](https://github.com/dandavison/delta) git pager:

`.zshenv`:

```zsh
# colors, smart-case, scrolloff (.5 to center), go to top
export LESS="-R -i -j3 +g"
# only "-R" is required
export LESSHISTFILE=$XDG_STATE_HOME/less_history
export LESSHISTSIZE=100
```

`less-delta` dirty wrapper script:

```bash
#!/bin/bash

# exit if empty, always paginate otherwise

[ ! -t 0 ] && input=$(cat -)
[ ! -t 0 ] && [ -z "$input" ] && exit 1

# fix n\N nagivation from delta
navigation_pattern="\"^(commit|added:|removed:|renamed:|Δ|•|LESS_DELTA_DELETEME)"
gawk -i inplace -v pattern="$navigation_pattern" "/\.mark/ { print pattern }; {print}" "$LESSHISTFILE"

printf '%s' "$input" | less $@

# cleanup
sed -i "/DELETEME/d" "$LESSHISTFILE"
```

`gitconfig`:

```gitconfig
# vi: ft=gitconfig

[core]
  # requires LESS env var
  # https://github.com/dandavison/delta/issues/497#issuecomment-1095383988
  # https://git-scm.com/docs/git-config#Documentation/git-config.txt-corepager
  pager = delta

[interactive]
  diffFilter = delta --color-only --features=interactive

[delta]
  navigate = true # use n and N to move between diff section
  pager = less-delta # NOTE: IMPORTANT
# optional after this line
  diff-highlight = true
  # https://github.com/neovim/neovim/issues/34102
  # hyperlinks = true
  tabs = 2
  features = decorations

# https://raw.githubusercontent.com/dandavison/delta/main/themes.gitconfig
# delta --help
[delta "decorations"]
  syntax-theme = OneHalfDark
  right-arrow = "->"
  # file
  file-added-label = Δ [A]
  file-removed-label = Δ [D]
  file-modified-label = Δ [M]
  file-renamed-label = Δ [R]
  file-copied-label = Δ [C]
  file-style = gray bold reverse
  # hunk
  hunk-header-style = omit
  # diff
  minus-style = red
  minus-emph-style = bold black red
  plus-style = green
  plus-emph-style = bold black green
  whitespace-error-style = yellow reverse
  # line numbers
  line-numbers = true
  line-numbers-left-style = gray
  line-numbers-right-style = gray
  # commit hash
  commit-style = bold yellow
  # grep
  hunk-label = ""
  hunk-header-file-style = none
  grep-line-number-style = white
  grep-file-style = bold gray
  grep-match-word-style = bold white gray

[delta "sidy-by-side"]
  side-by-side = true
```

to use `less-delta` pager just for yay, rather than setting pager in gitconfig's `delta.pager` alias yay as `DELTA_PAGER=less-delta yay`.

**Comment by HaleTom** [comment]
@danielwerg thanks for sharing.  Before I go and try it, what behaviour occurs after your fix?

**Comment by Jguer** [comment]
This has been implemented in #2905 , unfortunately I forgot to change squashed message.
what's been implemented is actually collecting all the diffs and then paginating those using a pager outside of git diff. 

This is a behavior also used in paru.
Please give feedback by trying out `yay-git`
